test/test_filters.rb

```diff
@@ -849,6 +849,20 @@ def to_liquid
         assert_equal 2, @filter.where(array, "color", nil).length
       end
 
+      should "filter array and string objects appropriately" do
```

Comment:
Based on this test, the functionality feels too complicated to fix the bug. The most straightforward way I know to test this is to write something like:
```ruby
should "not throw exception when an item doesn't have the property" do
  array = [{}, { "color" => nil }, { "color" => "" }, { "color" => "text" }, { "foo" => "bar" }]
  assert_equal [{ "color" => "text" }], @filter.where(hash, "color", "text")
end
```

Also, filtering on sub properties, as in #6939 (`context.type`), isn't valid syntax. This feels like we need a better reproduction case.

Reply:
Your test-case passes on `master`. The issue outlined in #6939 is actually due to `#item_property` encountering a `nil` element.

Modifying your test-case to the following (will fail on current `master`:
```ruby
should "not throw exception when an item doesn't have the property" do
  array = [{}, { "color" => "text" }, { "foo" => "bar" }, nil]
  assert_equal [{ "color" => "text" }], @filter.where(array, "color", "text")
end
```
Another case that'll fail on `master`:
```ruby
should "not throw exception when an item doesn't have the property" do
  array = [{}, { "color" => "text" }, [], { "foo" => "bar" }]
  assert_equal [{ "color" => "text" }], @filter.where(array, "color", "text")
end
```